drivers: adc: microchip: Introduce G1 ADC Driver

ArunMCHP opened the pull request:
This pull request adds ADC driver support for the Microchip ADC peripheral. It introduces the ADC driver implementation and updates the devicetree for the SAM D5x/E5x series to include the required node and bindings.

ArunMCHP committed:
dts: arm: microchip: Introduce ADC G1 dts binding

Add the device tree node and the binding file for
microchip adc G1 Peripheral.

Signed-off-by: Arunprasath P <arunprasath.p@microchip.com>

ArunMCHP committed:
drivers: adc: microchip: Introduce G1 ADC Driver

Add a Zephyr ADC driver for the Microchip G1 ADC peripheral
with support for differential mode, multi-channel sequencing,
oversampling, and NVM-based factory calibration.

Signed-off-by: Arunprasath P <arunprasath.p@microchip.com>

ArunMCHP committed:
boards: microchip: sam_e54_xpro: Add ADC pinctrl and feature support

Add ADC0 and ADC1 pinctrl definitions and update the board metadata
to list ADC as a supported feature on the SAM E54 Xplained Pro board.

Signed-off-by: Arunprasath P <arunprasath.p@microchip.com>

nandojve commented on dts/bindings/adc/microchip,adc-g1.yaml:
```diff
@@ -0,0 +1,74 @@
+
+  vref-mv:
+    type: int
+    description: |
```
User `must` select this in the board.

```diff
  vref-mv:
    type: int
+    required: true
    description: |
```

nandojve commented on dts/arm/microchip/sam/sam_d5x_e5x/common/samd5xe5x.dtsi:
```diff
@@ -254,6 +254,38 @@
+			interrupts = <118 0>, <119 0>;
+			interrupt-names = "overrun", "resrdy";
+			/* 48 MHz GCLK / 4 prescaler = 12 MHz ADC clock. */
+			prescaler = <4>;
```
prescaler is application dependent.
here we add only values that should never be changed from the user in the board.
drop `prescaler` and `vref-mv`.

nandojve commented on drivers/adc/adc_mchp_g1.c:
```diff
@@ -0,0 +1,911 @@
+
+static int8_t adc_set_oversampling(adc_registers_t *adc_reg, uint8_t oversampling)
+{
+
```
useless space

nandojve commented on drivers/adc/adc_mchp_g1.c:
```diff
@@ -0,0 +1,911 @@
+
+	adc_interrupt_clear(ADC_REGS);
+	result = adc_get_conversion_result(ADC_REGS);
+	*dev_data->buffer++ = result;
```
This is hard to read. Let compiler optimize the code.
`	*dev_data->buffer++ = result;`

nandojve commented on drivers/adc/adc_mchp_g1.c:
```diff
@@ -0,0 +1,911 @@
+			       struct k_poll_signal *async)
+{
+	struct adc_mchp_dev_data *data = dev->data;
+
```
drop empty line

nandojve commented on drivers/adc/adc_mchp_g1.c:
```diff
@@ -0,0 +1,911 @@
+		/* Invalid ADC instance */
+		LOG_ERR("Invalid ADC index for calibration");
+		break;
+	}
```
Should we consider the fail ASAP here ?
BTW, how this **n** will be invalid if it came from devicetree `DT_INST_FOREACH_STATUS_OKAY`?

```c
if (not valid index) {
    LOG_ERR("Invalid ADC index for calibration");
    return;
}

ADC_CALIB_WRITE(sw0_word, n, BIASCOMP);
ADC_CALIB_WRITE(sw0_word, n, BIASR2R);
ADC_CALIB_WRITE(sw0_word, n, BIASREFBUF);
```

ArunMCHP committed:
tests: adc: microchip: Add sam_e54_xpro overlay for tests

Add board overlay file for sam_e54_xpro to enable adc
test cases to run on this board

Signed-off-by: Arunprasath P <arunprasath.p@microchip.com>

nandojve commented on drivers/adc/adc_mchp_g1.c:
```diff
@@ -0,0 +1,917 @@
+		uint8_t val = (((*sw0_word) & ADC##adc_id##_##field##_Msk) >>                      \
+			       ADC##adc_id##_##field##_Pos);                                       \
+		ADC_REGS->ADC_CALIB |= ADC_CALIB_##field(val);                                     \
+	} while (0)
```
Is the difference between one instance to the another just an offset and the positions are the same? If yes, you should add the offset in devicetree instead to use adc_id.

ArunMCHP commented on drivers/adc/adc_mchp_g1.c:
```diff
@@ -0,0 +1,917 @@
+/* Set ADC Calib register */
+#if defined(ADC_CALIB_BIASCOMP_Pos)
+	if (dev_cfg->adc_instance_id == 0) {
+		ADC_CALIB_WRITE_FIELD(0, BIASCOMP);
```
* The calibration values are stored in the NVM Software Calibration Area, which is not related to the actual ADC peripheral address.
* The bit positions for these calibration fields differ for each ADC instance.
* Therefore, we cannot use the peripheral offset from the devicetree to determine them.

<img width="836" height="729" alt="image" src="https://github.com/user-attachments/assets/aa24c150-30eb-4aed-a9fc-62102b633476" />

<img width="813" height="356" alt="image" src="https://github.com/user-attachments/assets/5c49977b-d94d-4c57-bf19-aa53b2f4c003" />

ArunMCHP replied:
We are trying to use same G1 driver for SAM D5X/E5X, PIC32CX 41/61 and PIC32CM JH.

ArunMCHP committed:
samples: adc: Add sam_e54_xpro overlay for adc_sequence

Add a devicetree overlay to run the adc_sequence sample on
SAM E54 Xplained Pro.

Signed-off-by: Arunprasath P <arunprasath.p@microchip.com>

ArunMCHP commented on dts/bindings/adc/microchip,adc-g1.yaml:
```diff
@@ -0,0 +1,85 @@
+      ADC reference voltage in millivolts.
+
+      This should match the reference selected in hardware or software config.
+
```
Removed these props. Thanks!
